File: ts/stack.ts

class _Node<T>
{
    data: T;
    next: _Node<T> | null;

    constructor(data: T)
    {
        this.data = data;
        this.next = null;
    }
}

class Stack
{
    top: _Node<any> | null;
    length: number;

    constructor()
    {
        this.top = null;
        this.length = 0;
    }

    push(value: any) : void
    {
        let newNode = new _Node(value);
        newNode.next = this.top;
        this.top = newNode;
        this.length++;
    }

    isEmpty() : boolean
    {
        return this.length ? false : true;
    }

    size() : number
    {
        return this.length;
    }

    pop() : any
    {
        if (this.isEmpty())
        {
            console.log("cannot pop empty stack");
            return null;
        }
        else
        {
            let tmp = this.top!.data;
            this.top = this.top!.next ?? null;
            this.length--;
            return tmp;
        }
    }

    peek() : any
    {
        if (this.isEmpty())
        {
            console.log("No peek in empty stack");
            return null;
        }
        else
        {
            return this.top!.data;
        }
    }

    print() : void
    {
        process.stdout.write("[ ");
        let node = this.top;
        while(node){
            process.stdout.write(`${node.data}${node.next ? ', ' : ' '}`);
            node = node.next;
        }
        process.stdout.write("]\n");
    }
}


let s = new Stack();

s.push(1);
s.push(2);
s.push(3);
s.push("as");
console.log(`the stack size: ${s.size()}`);
s.print();
console.log(`the stack popped: ${s.pop()}, and the size is: ${s.size()}`);
console.log(`the stack peek: ${s.peek()}`)

File: cpp/stack.cpp
#include <bits/stdc++.h>
using namespace std;

template <typename T>
class Node
{
    public:
    T data;
    Node<T>* next;

    Node(T data)
    {
        this->data = data;
        this->next = nullptr;
    }
};

template <typename T>
class Stack
{
    int length;

    public:
    Node<T>* head;

    Stack()
    {
        this->length = 0;
        this->head = nullptr;
    }

    void push(T data)
    {
        Node<T>* newNode = new Node<T>(data);
        newNode->next = this->head;
        this->head = newNode;
        this->length++;
    }

    void print()
    {
        Node<T>* curr = this->head;
        cout << "[ ";
        while (curr != nullptr)
        {
            cout << curr->data << (curr->next != nullptr ? ", ": " ");  
            curr = curr->next;      
        }
        cout << "]\n";
    }

    bool isEmpty()
    {
        return (this->length ? true : false);
    }

    int size()
    {
        return this->length;
    }

    T peek()
    {
        if(this->length > 0)
            return this->head->data;

        else
        {
            cout << "the stack is empty, no peek\n";
            return T();
        }

    }

    T pop()
    {
        if(this->length > 0)
        {
            Node<T>* tmp = this->head;
            T data = tmp->data;
            this->head = tmp->next;
            delete tmp;
            this->length--;
            return data;
        }
        else
        {
            cout << "cannot pop empty stack\n";
            return T();
        }
        
    }
};


int main()
{
    Stack<string> s;
    s.push("i");
    s.push("will");
    s.push("pop");
    s.push("this");    
    s.print();
    cout << "stack peek: " << s.peek() << endl;
    cout << "stack size: " << s.size() << endl;
    cout << "stack pop: " << s.pop() << endl;
    cout << "stack size: " << s.size() << endl;
    s.print();

}